common/headtracker/types/head.go

```diff
@@ -0,0 +1,19 @@
+package types
+
+import (
+	txmgrtypes "github.com/smartcontractkit/chainlink/v2/common/txmgr/types"
+	"github.com/smartcontractkit/chainlink/v2/common/types"
+)
+
+//go:generate mockery --quiet --name Head --output ./mocks/ --case=underscore
+type Head[H types.Head[BLOCK_HASH], BLOCK_HASH types.Hashable, CHAIN_ID txmgrtypes.ID] interface {
```

Comment:
Is it possible to instead reuse the types.Head?
If you want additional functions, like Equals(), you can add them to that interface.

I didn't quite get why we need the other functions:
ChainID(), HasChainID() and IsValid()

Reply:
The main issue with reusing type head is the self referential problem that would arise when we are trying to compare pointer type in the method parameter. Without creating another Head, we would only be able to compare with a generic type.

Reply:
`ChainID()` is needed to retrieve the Head ID `EVMChainID` which is different from the Client's `ConfiguredID()`. 

The functions are used in the comparison line for checking if the chain id of the head is same as chainId of the client. 

https://github.com/smartcontractkit/chainlink/blob/d689da5ce76c11f504d68880404e9a5e3a0c3c72/core/chains/evm/headtracker/head_listener.go#L164
```
if !blockHeader.HasChainID() || blockHeader.ChainID().String() != chainId.String() {
```

Reply:
The reason why these functions are created is because we are not able to assign the methods to nil type like previously it was a pointer type instead of a generic type

Previously
```
			if blockHeader.EVMChainID == nil || !utils.NewBig(chainId).Equal(blockHeader.EVMChainID) {
				hl.logger.Panicf("head listener for %s received block header for %s", chainId, blockHeader.EVMChainID)
				```

Reply:
For ChainID() and other functions, that seems ok to add.

For Equals(), I don't get why we can't do Equals(Head[...]).
The implementation will be responsible for actual comparison.
So here, in EVM it will be like Equals(hh *evmtypes.head) bool

Reply:
If we do `Equals(Head[...])` on `common/types/head`, Our Equals function would have to be implemented like `Equals(r commontypes.Head[common.Hash])` instead of `Equals(r *Head)`. 

We would also get the following error
```
*"github.com/smartcontractkit/chainlink/v2/core/chains/evm/types".Head does not satisfy "github.com/smartcontractkit/chainlink/v2/common/types".Head["github.com/ethereum/go-ethereum/common".Hash] (wrong type for method Equals)
		have Equals(*"github.com/smartcontractkit/chainlink/v2/core/chains/evm/types".Head) bool
		want Equals("github.com/smartcontractkit/chainlink/v2/common/types".Head["github.com/ethereum/go-ethereum/common".Hash]) bool
```

I think that its because we would only be able to compare with a generic type, instead of pointer type Head.